File: dotnet/UrlShortener.Core/Services/UrlService.cs
using Microsoft.Extensions.Configuration;
using UrlShortener.Core.Interfaces;
using UrlShortener.Core.Models;

namespace UrlShortener.Core.Services;

public class UrlService : IUrlService
{
    private readonly IUrlDatabase _database;
    private readonly IConfiguration _configuration;

    public UrlService(IUrlDatabase database, IConfiguration configuration)
    {
        _database = database;
        _configuration = configuration;
    }

    public async Task<CreateUrlResponse> CreateShortUrlAsync(string originalUrl)
    {
        try
        {
            if (string.IsNullOrWhiteSpace(originalUrl) || !IsValidUrl(originalUrl))
            {
                return new CreateUrlResponse { Success = false, Error = "Invalid URL provided" };
            }

            var shortCode = GenerateShortCode();
            var attempts = 0;
            const int maxAttempts = 5;

            // Ensure the short code is unique
            while (await _database.UrlExistsAsync(shortCode) && attempts < maxAttempts)
            {
                shortCode = GenerateShortCode();
                attempts++;
            }

            if (attempts >= maxAttempts)
            {
                return new CreateUrlResponse { Success = false, Error = "Unable to generate unique short code" };
            }

            var result = await _database.CreateUrlAsync(shortCode, originalUrl);
            var baseUrl = _configuration["BaseUrl"] ?? "http://localhost:5000";

            return new CreateUrlResponse
            {
                Success = true,
                ShortCode = result.ShortCode,
                OriginalUrl = result.OriginalUrl,
                ShortUrl = $"{baseUrl}/api/{result.ShortCode}"
            };
        }
        catch (Exception ex)
        {
            return new CreateUrlResponse { Success = false, Error = $"Failed to create short URL: {ex.Message}" };
        }
    }

    public async Task<GetUrlResponse> GetOriginalUrlAsync(string shortCode)
    {
        try
        {
            if (string.IsNullOrWhiteSpace(shortCode))
            {
                return new GetUrlResponse { Success = false, Error = "Short code is required" };
            }

            var url = await _database.GetUrlAsync(shortCode);
            if (url == null)
            {
                return new GetUrlResponse { Success = false, Error = "Short URL not found" };
            }

            return new GetUrlResponse
            {
                Success = true,
                ShortCode = url.ShortCode,
                OriginalUrl = url.OriginalUrl,
                CreatedAt = url.CreatedAt
            };
        }
        catch (Exception ex)
        {
            return new GetUrlResponse { Success = false, Error = $"Failed to get original URL: {ex.Message}" };
        }
    }

    public async Task<GetAllUrlsResponse> GetAllUrlsAsync()
    {
        try
        {
            var urls = await _database.GetAllUrlsAsync();
            var baseUrl = _configuration["BaseUrl"] ?? "http://localhost:5000";

            var urlItems = urls.Select(url => new UrlItem
            {
                ShortCode = url.ShortCode,
                OriginalUrl = url.OriginalUrl,
                ShortUrl = $"{baseUrl}/api/{url.ShortCode}",
                CreatedAt = url.CreatedAt
            }).ToList();

            return new GetAllUrlsResponse
            {
                Success = true,
                Urls = urlItems
            };
        }
        catch (Exception ex)
        {
            return new GetAllUrlsResponse { Success = false, Error = $"Failed to get all URLs: {ex.Message}" };
        }
    }

    private static string GenerateShortCode()
    {
        const string chars = "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789";
        var random = new Random();
        return new string(Enumerable.Repeat(chars, 4)
            .Select(s => s[random.Next(s.Length)]).ToArray());
    }

    private static bool IsValidUrl(string url)
    {
        return Uri.TryCreate(url, UriKind.Absolute, out _);
    }
}

File: node/function/src/UrlService.js
const shortid = require('shortid');

class UrlService {
  constructor(database) {
    this.database = database;
  }

  // Generate a short code (4 characters alphanumeric)
  generateShortCode() {
    const chars = 'ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789';
    let result = '';
    for (let i = 0; i < 4; i++) {
      result += chars.charAt(Math.floor(Math.random() * chars.length));
    }
    return result;
  }

  // Validate URL format
  isValidUrl(url) {
    try {
      new URL(url);
      return true;
    } catch {
      return false;
    }
  }

  // Create a shortened URL
  async createShortUrl(originalUrl) {
    if (!originalUrl || !this.isValidUrl(originalUrl)) {
      throw new Error('Invalid URL provided');
    }

    let shortCode = this.generateShortCode();
    let attempts = 0;
    const maxAttempts = 5;

    // Ensure the short code is unique
    while (await this.database.urlExists(shortCode) && attempts < maxAttempts) {
      shortCode = this.generateShortCode();
      attempts++;
    }

    if (attempts >= maxAttempts) {
      throw new Error('Unable to generate unique short code');
    }

    try {
      const result = await this.database.createUrl(shortCode, originalUrl);
      return {
        success: true,
        shortCode: result.shortCode,
        originalUrl: result.originalUrl,
        id: result.id
      };
    } catch (error) {
      throw new Error(`Failed to create short URL: ${error.message}`);
    }
  }

  // Get original URL by short code
  async getOriginalUrl(shortCode) {
    if (!shortCode) {
      throw new Error('Short code is required');
    }

    try {
      const url = await this.database.getUrl(shortCode);
      if (!url) {
        return { success: false, error: 'Short URL not found' };
      }

      return {
        success: true,
        shortCode: url.shortCode,
        originalUrl: url.originalUrl,
        createdAt: url.createdAt
      };
    } catch (error) {
      throw new Error(`Failed to get original URL: ${error.message}`);
    }
  }

  // Get all URLs
  async getAllUrls() {
    try {
      const urls = await this.database.getAllUrls();
      // Do not include an absolute shortUrl here; return the shortCode and
      // let the frontend compute the final URL using its current origin.
      return {
        success: true,
        urls: urls.map(url => ({
          shortCode: url.shortCode,
          originalUrl: url.originalUrl,
          createdAt: url.createdAt
        }))
      };
    } catch (error) {
      throw new Error(`Failed to get all URLs: ${error.message}`);
    }
  }
}

module.exports = UrlService;
